src/sdk/namespace_fs.js

```diff
@@ -492,24 +492,93 @@ class NamespaceFS {
+        } else {
+            try {
+                await this._upload_stream(params.source_stream, upload_path, fs_account_config);
+                // TODO use file xattr to store md5_b64 xattr, etc.
+                const stat = await nb_native().fs.stat(fs_account_config, upload_path);
+                await nb_native().fs.rename(fs_account_config, upload_path, file_path);
+                return { etag: this._get_etag(stat) };
+            } catch (err) {
+                throw this._translate_object_error_codes(err);
+            }
+        }
+    }
 
-            await Promise.all([this._make_path_dirs(file_path, fs_account_config), this._make_path_dirs(upload_path, fs_account_config)]);
-            await this._upload_stream(params.source_stream, upload_path, fs_account_config);
-            // TODO use file xattr to store md5_b64 xattr, etc.
-            const stat = await nb_native().fs.stat(fs_account_config, upload_path);
-            await nb_native().fs.rename(fs_account_config, upload_path, file_path);
-            return { etag: this._get_etag(stat) };
-        } catch (err) {
-            throw this._translate_object_error_codes(err);
+    async _copy_stream(source_file_path, upload_path, fs_account_config) {
+        let target_file;
+        let source_file;
+        try {
+            //Opening the source and target files
+            source_file = await nb_native().fs.open(fs_account_config, source_file_path);
+            target_file = await nb_native().fs.open(fs_account_config, upload_path, 'w');
+            //Reading the source_file and writing into the target_file
+            const { buffer } = await buffers_pool.get_buffer(config.NSFS_BUF_SIZE);
+            let read_pos = 0;
+            for (;;) {
+                const bytesRead = await source_file.read(fs_account_config, buffer, 0, config.NSFS_BUF_SIZE, read_pos);
+                if (!bytesRead) break;
+                read_pos += bytesRead;
+                const data = buffer.slice(0, bytesRead);
+                await target_file.write(fs_account_config, data);
+            }
+            //Closing the source_file and target files
+            await source_file.close(fs_account_config);
+            source_file = null;
+            await target_file.close(fs_account_config);
+            target_file = null;
+        } catch (e) {
+            console.error('Failed to copy object', e);
+        } finally {
+            try {
+                if (source_file) await source_file.close(fs_account_config);
+                if (target_file) await target_file.close(fs_account_config);
+            } catch (err) {
+                console.warn('NamespaceFS: upload_object - copy_source file close error', err);
+            }
+        }
+    }
+
+    //Checking if we can creat a link 
+    async _can_create_link(other) { //LMLM what is other? 
+        //Add here more conditions
+        if (this.is_same_namespace && other.fs_backend === this.fs_backend) {
```

Comment:
did you mean to call is_same_namespace()?

Reply:
Also why isn't the compare of fs_backend added into is_same_namespace?

Reply:
@guymguym you are correct, 
I wanted this function at first to add some checks like `File exists` or any other that we will think of.
I could think of a few:
- is same namespace types (which are and should be in `is_same_namespace`)
- are POSIX calls, such as `File exists` and any other restrictions that prevent a link (maybe "No such file or directory"), but calling the ops might be expensive so we should just try to do link and failback to normal copy if we are not succeeding.
- Copy requests such as  MetadataDirective 'COPY' or maybe TaggingDirective 'COPY'.

I think that the `_can_create_link` function is not needed as we fall back to copy on fail.